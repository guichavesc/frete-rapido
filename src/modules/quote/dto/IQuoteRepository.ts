export interface IQuoteRepository {
  save: (quote: {
    name: string;
    id?: string;
    service: string;
    deadline: string;
    price: number;
  }) => Promise<void>;

  saveMany(
    quotes: {
      id?: string;
      name: string;
      service: string;
      deadline: string;
      price: number;
    }[]
  ): Promise<void>;

  list(lastQuotes?: number): Promise<
    {
      id: string;
      name: string;
      service: string;
      deadline: string;
      price: number;
    }[]
  >;
}
