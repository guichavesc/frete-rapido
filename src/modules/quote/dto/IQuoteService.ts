export type TCalculateRequest = {
  recipient: {
    address: {
      zipcode: string;
    };
  };
  volumes: {
    category: number;
    amount: number;
    unitary_weight: number;
    price: number;
    sku: string;
    height: number;
    width: number;
    length: number;
  }[];
};

export interface IQuoteService {
  calculate(data: TCalculateRequest): Promise<TCalculateResponse>;
}

export interface TCalculateResponse {
  dispatchers: Dispatcher[];
}

export interface Dispatcher {
  id: string;
  request_id: string;
  registered_number_shipper: string;
  registered_number_dispatcher: string;
  zipcode_origin: number;
  offers: Offer[];
  total_price: number;
}

export interface Offer {
  offer: number;
  table_reference: string;
  simulation_type: number;
  carrier: Carrier;
  service: string;
  delivery_time: DeliveryTime;
  expiration: string;
  cost_price: number;
  final_price: number;
  weights: Weights;
  original_delivery_time: OriginalDeliveryTime;
  correios?: Correios;
  service_code?: string;
  service_description?: string;
}

export interface Carrier {
  name: string;
  registered_number: string;
  state_inscription: string;
  logo: string;
  reference: number;
  company_name: string;
}

export interface DeliveryTime {
  days: number;
  estimated_date: string;
  hours?: number;
  minutes?: number;
}

export interface Weights {
  real: number;
  cubed?: number;
  used: number;
}

export interface OriginalDeliveryTime {
  days: number;
  estimated_date: string;
  hours?: number;
  minutes?: number;
}

export interface Correios {}
