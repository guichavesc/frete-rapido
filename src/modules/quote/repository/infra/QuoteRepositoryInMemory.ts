import { v4 as uuid } from "uuid";
import { IQuoteRepository } from "../../dto/IQuoteRepository";
export class QuoteRepositoryInMemory implements IQuoteRepository {
  private quotes = [];

  async save(quote: {
    id?: string;
    name: string;
    service: string;
    deadline: string;
    price: number;
  }) {
    const quoteData = {
      ...(quote.id && {
        id: uuid(),
      }),
      name: quote.name,
      service: quote.service,
      deadline: quote.deadline,
      price: quote.price,
      createdAt: "2000-01-01T00:00:00.000Z",
    };

    this.quotes.push(quoteData);
  }

  async saveMany(
    quotes: {
      id?: string;
      name: string;
      service: string;
      deadline: string;
      price: number;
    }[]
  ) {
    quotes.forEach(async (quote) => {
      await this.save(quote);
    });
  }

  async list(lastQuotes?: number) {
    if (lastQuotes) {
      return this.quotes.slice(-lastQuotes);
    }

    return this.quotes;
  }
}
