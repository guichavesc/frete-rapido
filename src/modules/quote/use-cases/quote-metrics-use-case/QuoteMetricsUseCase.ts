import { IQuoteRepository } from "../../dto/IQuoteRepository";

export class QuoteMetricsUseCase {
  constructor(private readonly quoteRepository: IQuoteRepository) {}
  async execute({ lastQuotes }) {
    const quotes = await this.quoteRepository.list(lastQuotes);

    const totalQuotesPerCarrier = quotes.reduce((acc, quote) => {
      if (!acc[quote.name]) {
        acc[quote.name] = 0;
      }

      acc[quote.name] += 1;

      return acc;
    }, {});

    const priceAveragePerCarrier = quotes.reduce((acc, quote) => {
      if (!acc[quote.name]) {
        acc[quote.name] = 0;
      }

      acc[quote.name] = +(
        (quote.price + acc[quote.name]) /
        totalQuotesPerCarrier[quote.name]
      ).toFixed(2);

      return acc;
    }, {});

    const totalPricePerCarrier = quotes.reduce((acc, quote) => {
      if (!acc[quote.name]) {
        acc[quote.name] = 0;
      }

      acc[quote.name] = +(quote.price + acc[quote.name]).toFixed(2);

      return acc;
    }, {});

    const cheapestPricePerCarrier =
      quotes?.length > 0
        ? quotes.reduce(function (res, obj) {
            return obj.price < res.price ? obj : res;
          })
        : null;

    const expensivePricePerCarrier =
      quotes?.length > 0
        ? quotes.reduce(function (res, obj) {
            return obj.price > res.price ? obj : res;
          })
        : null;

    return {
      metrics: {
        totalQuotes: quotes.length,
        totalQuotesPerCarrier,
        totalPricePerCarrier,
        priceAveragePerCarrier,
        cheapestPricePerCarrier,
        expensivePricePerCarrier,
      },
    };
  }
}
