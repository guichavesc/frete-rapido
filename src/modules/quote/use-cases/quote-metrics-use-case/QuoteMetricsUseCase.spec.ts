import { IQuoteRepository } from "../../dto/IQuoteRepository";
import { QuoteRepositoryInMemory } from "../../repository/infra/QuoteRepositoryInMemory";
import { QuoteMetricsUseCase } from "./QuoteMetricsUseCase";

let quoteMetricsUseCase: QuoteMetricsUseCase;
let quoteRepository: IQuoteRepository;

const quotes = [
  {
    name: "Correios",
    deadline: "5",
    price: 5.5,
    service: "SEDEX",
  },
  {
    name: "Braspress",
    deadline: "5",
    price: 12.5,
    service: "Expresso",
  },
  {
    name: "Braspress",
    deadline: "5",
    price: 35,
    service: "Normal",
  },
  {
    name: "Jadlog",
    deadline: "5",
    price: 12.5,
    service: "Expresso",
  },
  {
    name: "Jadlog",
    deadline: "5",
    price: 3.5,
    service: "Normal",
  },
  {
    name: "Jadlog",
    deadline: "5",
    price: 32.5,
    service: "Normal 5 days",
  },
];
describe("Quote Metrics Use Case", () => {
  beforeEach(() => {
    quoteRepository = new QuoteRepositoryInMemory();
    quoteMetricsUseCase = new QuoteMetricsUseCase(quoteRepository);
  });

  it("Should return the metrics", async () => {
    await quoteRepository.saveMany(quotes);

    const data = await quoteMetricsUseCase.execute({
      lastQuotes: 10,
    });

    expect(data).toMatchObject({
      metrics: {
        totalQuotes: 6,
        totalQuotesPerCarrier: {
          Correios: 1,
          Braspress: 2,
          Jadlog: 3,
        },
        totalPricePerCarrier: {
          Correios: 5.5,
          Braspress: 47.5,
          Jadlog: 48.5,
        },
        priceAveragePerCarrier: {
          Correios: 5.5,
          Braspress: 20.63,
          Jadlog: 11.69,
        },
        cheapestPricePerCarrier: {
          name: "Jadlog",
          deadline: "5",
          price: 3.5,
          service: "Normal",
          // jest should ignore date
          createdAt: "2000-01-01T00:00:00.000Z",
        },
        expensivePricePerCarrier: {
          name: "Braspress",
          deadline: "5",
          price: 35,
          service: "Normal",
          createdAt: "2000-01-01T00:00:00.000Z",
        },
      },
    });
  });

  it("Should return the metrics with last 3 quotes", async () => {
    await quoteRepository.saveMany(quotes);

    const data = await quoteMetricsUseCase.execute({
      lastQuotes: 3,
    });

    expect(data).toMatchObject({
      metrics: {
        totalQuotes: 3,
        totalQuotesPerCarrier: {
          Jadlog: 3,
        },
        totalPricePerCarrier: {
          Jadlog: 48.5,
        },
        priceAveragePerCarrier: {
          Jadlog: 11.69,
        },
        cheapestPricePerCarrier: {
          name: "Jadlog",
          deadline: "5",
          price: 3.5,
          service: "Normal",
          // jest should ignore date
          createdAt: "2000-01-01T00:00:00.000Z",
        },
        expensivePricePerCarrier: {
          name: "Jadlog",
          deadline: "5",
          price: 32.5,
          service: "Normal 5 days",
          createdAt: "2000-01-01T00:00:00.000Z",
        },
      },
    });
  });

  it("Should return the metrics with last 1 quotes", async () => {
    await quoteRepository.saveMany(quotes);

    const data = await quoteMetricsUseCase.execute({
      lastQuotes: 1,
    });

    expect(data).toMatchObject({
      metrics: {
        totalQuotes: 1,
        totalQuotesPerCarrier: {
          Jadlog: 1,
        },
        totalPricePerCarrier: {
          Jadlog: 32.5,
        },
        priceAveragePerCarrier: {
          Jadlog: 32.5,
        },
        cheapestPricePerCarrier: {
          name: "Jadlog",
          deadline: "5",
          price: 32.5,
          service: "Normal 5 days",
          createdAt: "2000-01-01T00:00:00.000Z",
        },
        expensivePricePerCarrier: {
          name: "Jadlog",
          deadline: "5",
          price: 32.5,
          service: "Normal 5 days",
          createdAt: "2000-01-01T00:00:00.000Z",
        },
      },
    });
  });
});
