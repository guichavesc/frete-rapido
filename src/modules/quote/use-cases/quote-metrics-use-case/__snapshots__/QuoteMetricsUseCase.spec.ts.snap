// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote Metrics Use Case Should return the metrics  1`] = `
{
  "metrics": {
    "cheapestPricePerCarrier": {
      "createdAt": Any<Date>,
      "deadline": "5",
      "name": "Jadlog",
      "price": 3.5,
      "service": "Normal",
    },
    "expensivePricePerCarrier": {
      "createdAt": Any<Date>,
      "deadline": "5",
      "name": "Braspress",
      "price": 35,
      "service": "Normal",
    },
    "priceAveragePerCarrier": {
      "Braspress": 23.75,
      "Correios": 5.5,
      "Jadlog": 16.17,
    },
    "totalPricePerCarrier": {
      "Braspress": 47.5,
      "Correios": 5.5,
      "Jadlog": 48.5,
    },
    "totalQuotes": 6,
    "totalQuotesPerCarrier": {
      "Braspress": 2,
      "Correios": 1,
      "Jadlog": 3,
    },
  },
}
`;
